### PR TITLE
added DocumentManagerDecorator (BC break in several internal classes)

### PR DESCRIPTION
**Pull request by dbu**
@lsmith77  this is a quite old branch. does it make sense to keep and continue this refactoring, or should we delete it?

### CODE REVIEW

**Comment by lsmith77** [comment]
ORM seems to have this: http://www.doctrine-project.org/2013/09/11/doctrine-2-4-released.html
IMHO yeah we should wrap this up.

File: lib/Doctrine/ODM/PHPCR/ChildrenCollection.php [hunk 1]
```diff
@@ -42,13 +42,13 @@ class ChildrenCollection extends PersistentCollection
     /**
      * Creates a new persistent collection.
      *
-     * @param DocumentManager $dm         The DocumentManager the collection will be associated with.
-     * @param object          $document   The parent document instance
-     * @param string|array    $filter     Filter string or array of filter string
-     * @param int             $fetchDepth Optional fetch depth, -1 to not override.
-     * @param string          $locale     The locale to use during the loading of this collection
+     * @param DocumentManagerInterface $dm  The DocumentManager the collection will be associated with.
```
Comment (lsmith77) [inline lib/Doctrine/ODM/PHPCR/ChildrenCollection.php]: i am not a huge fan of aligning in general (as it pollutes diffs), but i think re-aligning everything to be perfect here is imho really ugly

File: lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+use Doctrine\ODM\PHPCR\Translation\TranslationStrategy\TranslationStrategyInterface;
+use PHPCR\PropertyType;
+use PHPCR\Query\QueryInterface;
+
+/**
+ * Base class for DocumentManager decorators
+ *
+ * @since   1.3
+ */
+abstract class DocumentManagerDecorator extends ObjectManagerDecorator implements DocumentManagerInterface
```
Comment (lsmith77) [inline lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php]: should we also decorate DocumentManager:create() ?

File: tests/Doctrine/Tests/ODM/PHPCR/Decorator/DocumentManagerDecoratorTest.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+namespace Doctrine\Tests\ODM\PHPCR\Decorator;
+
+use Doctrine\Tests\ODM\PHPCR\PHPCRTestCase;
+
+/**
+ * @group unit
+ */
+class DocumentManagerDecoratorTest extends PHPCRTestCase
+{
+    public function testCheckIfAllPublicMethodsAreDecorated()
```
Comment (lsmith77) [inline tests/Doctrine/Tests/ODM/PHPCR/Decorator/DocumentManagerDecoratorTest.php]: just a simple test to check if we decorate all the relevant methods.

**Comment by lsmith77** [comment]
updated the PR

File: lib/Doctrine/ODM/PHPCR/DocumentManagerInterface.php [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+     * Note that this method only returns referrers that have been flushed.
+     *
+     * @param object       $document the target of the references to be loaded
+     * @param string|null  $type     the reference type, null|'weak'|'hard'
+     * @param string|null  $name     optional PHPCR property name that holds
+     *                               the reference
+     * @param string       $locale   the locale to use during the loading of
+     *                               this collection
+     * @param string|null  $refClass Class the referrer document must be
+    instanceof.
```
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/DocumentManagerInterface.php]: something is off here. tabs?

File: lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php (hunk 2 above)
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php]: i would not decorate that. its a static method anyways. i would assume that at the moment of instantiation, you know what you want to instantiate...

Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php]: orm also does not do it. i would leave that be.

File: lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+use Doctrine\ODM\PHPCR\DocumentManagerInterface;
+use Doctrine\ODM\PHPCR\Translation\LocaleChooser\LocaleChooserInterface;
+use Doctrine\ODM\PHPCR\Translation\TranslationStrategy\TranslationStrategyInterface;
+use PHPCR\PropertyType;
+use PHPCR\Query\QueryInterface;
+
+/**
+ * Base class for DocumentManager decorators
+ *
+ * @since   1.3
```
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Decorator/DocumentManagerDecorator.php]: too many spaces.

File: lib/Doctrine/ODM/PHPCR/DocumentManagerInterface.php (hunk 4 above)
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/DocumentManagerInterface.php]: Missing `    * ` prefix I think.

**Comment by dbu** [comment]
i will quickly fix these issues, add a changelog entry and merge.

**Pull request merged**